src/util/address.rs

```diff
@@ -135,7 +135,189 @@ impl From<bech32::Error> for Error {
     }
 }
 
-/// The different types of addresses.
+/// Defines which witness version may have an address or transaction output.
+///
+/// Represents classification of transaction output ScriptPubkey which can be extracted from either
+/// [`TxOut`] data or any given [`Address`] even for non-spent outputs and without the need for a
+/// blockchain access.
+///
+/// The structure is required to support some ambiguity on the witness version used by some address,
+/// since `Option<`[`WitnessVersion`]`>` can't cover that ambiguity (see details in
+/// [`SegWitInfo::Ambiguous`] description).
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum SegWitInfo {
+    /// P2PKH addresses
+    PreSegWit,
```

Comment:
`Option::None` would feel more idiomatic, do you have a particular reason to not use it?

Reply:
Yes: ambiguity with P2SH (which may hide P2WSH/WPKH-in-P2SH), as explained in comments below. So `Option<SegWitInfo>` does not work here, we need three alternatives like "non-segwit", "maybe segwit" and "segwit"

Reply:
`None` being "non-segwit" would feel natural to me but I can understand that you can treat them as being on same level.

Reply:
I would prefer adding APIs like `witness_version` directly to `AddressInfo`. The struct `SegwitInfo` has only one function of use that provides the witness version of the address. I don't see why we need 3 constructors for it and this amount of code bloat for it.

Reply:
@sanket1729 `SegWitInfo` is needed since it is returned not only by address types, but also from Script in later commits - and you cant use `AddressInfo` for that